server/src/main/java/org/opensearch/action/admin/indices/shrink/TransportResizeAction.java

```diff
@@ -250,6 +258,39 @@ static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(
             .copySettings(resizeRequest.getCopySettings() == null ? false : resizeRequest.getCopySettings());
     }
 
+    /**
+     * Calculate target index's shards number according to maxShardSize and the source index's storage(only primary shards included)
+     * @param maxShardSize the maximum size of a primary shard in the target index
+     * @param sourceIndexShardStoreStats primary shards' store stats of the source index
+     * @param sourceIndexMetaData source index's metadata
+     * @return target index's shards number
+     */
+    protected static int calculateTargetIndexShardsNum(
+        ByteSizeValue maxShardSize,
+        StoreStats sourceIndexShardStoreStats,
+        IndexMetadata sourceIndexMetaData
+    ) {
+        if (maxShardSize == null
+            || sourceIndexShardStoreStats == null
+            || maxShardSize.getBytes() == 0
+            || sourceIndexShardStoreStats.getSizeInBytes() == 0) {
+            return 1;
+        }
+
+        int sourceIndexShardsNum = sourceIndexMetaData.getNumberOfShards();
+        int minValue = (int) Math.ceil((double) sourceIndexShardStoreStats.getSizeInBytes() / maxShardSize.getBytes());
+        if (minValue >= sourceIndexShardsNum) {
+            return sourceIndexShardsNum;
+        }
+
+        for (int i = minValue; i < sourceIndexShardsNum; i++) {
```

Comment:
It seems like we calculate the factor here, why?

Reply:
The `minValue` may not be a factor of the source index's shards count, so we should iterate over all the numbers between `minValue` and `sourceIndexShardsNum` to find the lowest factor.

Reply:
@gaobinlong why do we need the factor value?

Reply:
That's because if we use the factor value, we don't need to rehash documents after the new shrunken index is created  and data is still balanced between the new shards.  According to shrink's implementation, when the new shrunken index is created, we will hard-links the segment files from the source index into the new index(if the file system supports hard-link), each new shard will be hard-linked to equal amount of source shards, for example, the source index has 6 primary shards while the new shrunken index has 2 primary shards, shard[0] in the new shrunken index will be hard-linked to shard[0],shard[1],shard[2] in the source index, shard[1] in the new shrunken index will be hard-linked to shard[3],shard[4],shard[5], each new shard has equal amount of documents and routing from document to shard will not be changed, so we don't need to rehash documents.

Reply:
Makes perfect sense, thank you